tensorflow/core/kernels/svd_op_gpu.cu.cc

```diff
@@ -0,0 +1,249 @@
+
+    auto input_reshaped = inputCopy.template flat_inner_dims<Scalar, 3>();
+
+    // Reshape output tensors
+    auto outputS_reshaped = outputS->template flat_inner_dims<Scalar, 2>();
+    Scalar* outputU_reshaped_ptr = NULL;
+    Scalar* outputVT_reshaped_ptr = NULL;
+    if (compute_uv_) {
+      auto outputU_reshaped = outputU->template flat_inner_dims<Scalar, 3>();
+      auto outputVT_reshaped = outputVT.template flat_inner_dims<Scalar, 3>();
+      outputU_reshaped_ptr = outputU_reshaped.data();
+      outputVT_reshaped_ptr = outputVT_reshaped.data();
+    }
+
+    // Launch a SVD kernel for each matrix in the batch.
+    const int64 batch_size = input_reshaped.dimension(0);
+    DeviceLapackInfo dev_info(context, batch_size, "gesvd");
+
+    // TODO(rmlarsen): Parallelize over batches if it turns out to be
+    // an important use case.
+    CudaSolver solver(context);
+    for (int64 i = 0; i < batch_size; ++i) {
+      Scalar* input_ptr = input_reshaped.data() + i * m * n;
+      Scalar* outputS_ptr = outputS_reshaped.data() + i * p;
+      int lda = m;
+
+      Scalar* outputU_ptr = NULL;
+      Scalar* outputVT_ptr = NULL;
+      signed char jobu = 'N';
+      signed char jobvt = 'N';
+      int ldu = m;
+      int ldvt = n;
+      if (compute_uv_) {
+        if (full_matrices_) {
+          outputU_ptr = outputU_reshaped_ptr + i * m * m;
+          outputVT_ptr = outputVT_reshaped_ptr + i * n * n;
+          jobu = 'A';
+          jobvt = 'A';
+        } else {
+          outputU_ptr = outputU_reshaped_ptr + i * m * p;
+          outputVT_ptr = outputVT_reshaped_ptr + i * n * p;
+          jobu = 'S';
+          jobvt = 'S';
+        }
+      }
+
+      int* dev_info_ptr = dev_info.mutable_data() + i;
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          solver.Gesvd(jobu, jobvt, m, n, input_ptr, lda, outputS_ptr,
+                       outputU_ptr, ldu, outputVT_ptr, ldvt, dev_info_ptr),
+          done);
+    }
+
+    // Test if it was successfull.
+    // I'm not using solver.CopyLapackInfoToHostAsync
+    // because it resulted in a memory corruption on the host
+    // (because I have some operations afterwards, so I can't use done())
+    HostLapackInfo host_info(context, batch_size, "gesvd");
+    cudaMemcpy(host_info.mutable_data(), dev_info.data(),
+               sizeof(int) * batch_size, cudaMemcpyDeviceToHost);
+    Status status;
+    for (int i = 0; i < batch_size && status.ok(); ++i) {
+      const int info_value = host_info.data()[i];
+      if (info_value != 0) {
+        status = errors::InvalidArgument(
+            "Got info = ", info_value, " for batch index ", i,
+            ", expected info = 0. Debug_info =", host_info.debug_info());
+      }
+    }
+    if (!status.ok()) {
+      status.Update(errors::InvalidArgument(kErrMsg));
+    }
+    OP_REQUIRES_OK_ASYNC(context, status, done);
+    // TODO: Maybe switch to solver.CopyLapackInfoToHostAsync again
```

Comment:
Sort of, calling DoTranspose inside the closure is the trick you are looking for.